src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,286 @@
+/** @jsx React.DOM */
+
+var React = require('react');
+var Util = require("./util.js");
+
+var SearchAndReplaceDialog = React.createClass({
+
+    getDefaultProps: function() {
+        return {
+            onChange() {}    // TODO: throw an error
+        };
+    },
+
+    getInitialState() {
+        return {
+            replaceString: "",
+            searchResultCount: 0
+        }
+    },
+
+    updateSearchString(event) {
+        var searchIndex = 0;
+        var searchString = event.target.value;
+        var searchResultCount = this.getSearchResultCount(searchString);
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ searchString, searchIndex });
+    },
+
+    updateReplaceString(event) {
+        this.setState({ replaceString: event.target.value });
+    },
+
+    updateSearchResults() {
+        var searchIndex = 0;
+        var searchResultCount = this.getSearchResultCount(this.props.searchString);
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ searchIndex });
+    },
+
+    getSearchResultCount(searchString) {
+        if (searchString === "") {
+            return 0;
+        }
+
+        var count = 0;
+
+        if (this.props.question) {
+            count += Util.countOccurrences(this.props.question.content, searchString);
+        }
+
+        if (this.props.hints) {
+            this.props.hints.forEach(hint => {
+                count += Util.countOccurrences(hint.content, searchString);
+            });
+        }
+
+        if (this.props.particle) {
+            var particle = this.props.particle;
+            if (Array.isArray(particle)) {
```

Comment:
We should probably just pass in `this._sections()` from ArticleEditor here, and then we can know that it's always an array.